contracts/Market.sol

```diff
@@ -0,0 +1,857 @@
+      recipient,
+      numberOfRemovals,
+      ids,
+      amounts,
+      suppliers
+    );
+  }
+
+  /**
+   * @dev The number of distinct removal token ids listed in the market that are not reserved.
+   * todo rm and calc off-chain if needed from this.numberOfReservedRemovals() - this.reservedSupply().length
+   */
+  function numberOfUnreservedRemovals() external view returns (uint256) {
+    return
+      _removal.numberOfTokensOwnedByAddress(address(this)) -
+      this.numberOfReservedRemovals(); // todo gas vs _reservedSupply.length() ?
+  }
+
+  /**
+   * @dev The number of distinct removal token ids listed in the market that are reserved.
+   * todo rm and calc off-chain if needed from this.reservedSupply().length
+   */
+  function numberOfReservedRemovals() external view returns (uint256) {
+    return _reservedSupply.length();
+  }
+
+  /**
+   * @dev The distinct removal token ids listed in the market that are reserved.
+   */
+  function reservedSupply() external view returns (uint256[] memory) {
+    return _reservedSupply.values();
+  }
+
+  /**
+   * @dev Reverts if market is out of stock or if available stock is being reserved for priority buyers
+   * and buyer is not priority.
+   *
+   * @param purchaseAmount The number of carbon removals being purchased.
+   */
+  function _checkSupply(uint256 purchaseAmount) private view {
+    // TODO: BUG: when using swap from single supplier, this function should check against the suppliers active balance,
+    //not the markets total active balance!
+    // TODO: BUG: this function doesn't consider reserved removals!
+    uint256 activeSupply = _removal.cumulativeBalanceOf(address(this));
+    if (activeSupply == 0) {
+      revert OutOfStock();
+    }
+    if (activeSupply <= _priorityRestrictedThreshold) {
+      if (!hasRole(ALLOWLIST_ROLE, _msgSender())) {
+        revert LowSupplyAllowlistRequired();
+      }
+    }
+    if (purchaseAmount > activeSupply) {
+      revert InsufficientSupply(); // todo test the logic for these checks is correct for all states
+    }
+  }
+
+  /**
+   * @dev Determines the removal ids, amounts, and suppliers to fill the given purchase quantity in
+   * a round-robin order.
+   *
+   * @param certificateAmount The number of carbon removals to purchase.
+   * @return numberOfRemovalForOrder The number of distinct removal token ids used to fulfill this order.
+   * @return ids An array of the removal token ids being drawn from to fulfill this order.
+   * @return amounts An array of amounts being allocated from each corresponding removal token.
+   * @return suppliers The address of the supplier who owns each corresponding removal token.
+   */
+  function _allocateSupplyRoundRobin(uint256 certificateAmount)
+    private
+    returns (
+      uint256,
+      uint256[] memory,
+      uint256[] memory,
+      address[] memory
+    )
+  {
+    uint256 remainingAmountToFill = certificateAmount;
+    uint256 numberOfActiveRemovalsInMarket = _removal
```

Comment:
we have a view function that does this, can we use that?

Reply:
view function removed in #300